cpp/src/lance/format/schema.cc

```diff
@@ -382,6 +382,48 @@ std::shared_ptr<Field> Field::Project(const std::shared_ptr<::arrow::Field>& arr
   return new_field;
 }
 
+::arrow::Result<std::shared_ptr<Field>> Field::Merge(const Field& other) const {
+  if (name_ != other.name_) {
+    return ::arrow::Status::Invalid(
+        fmt::format("Field::merge: attempt to merge two fields with different names: {} != {}",
+                    name_,
+                    other.name_));
+  };
+  if (logical_type_ != other.logical_type_) {
+    return ::arrow::Status::Invalid(
+        fmt::format("Field::merge: attempt to merge two fields with different types: {} != {}",
+                    logical_type_,
+                    other.logical_type_));
+  }
+  auto field = Copy();
+  auto dtype = type();
+
+  if (arrow::is_struct(dtype->id())) {
+    for (auto& child : children_) {
+      auto other_child = other.Get(child->name_);
+      if (other_child) {
+        // Both sides have the field, let's merge them.
+        ARROW_ASSIGN_OR_RAISE(auto merged, child->Merge(other));
+        field->AddChild(merged);
+      } else {
+        // Only I have the field.
+        field->AddChild(child->Copy(true));
+      }
+    }
+    for (auto& child : other.children_) {
+      if (Get(child->name_) == nullptr) {
+        // Have not seen this child before
+        field->AddChild(child->Copy(true));
+      }
+    }
+  } else if (::arrow::is_list_like(dtype->id())) {
```

Comment:
if one field is `fixed_sized_list<T, list_size=4>` and the other is `fixed_sized_list<T, list_size=2>`, I believe this would consider these two mergeable. Should it be?